### PR TITLE
fix #4703 menuinst PathNotFoundException

### LINKED ISSUE #4703: Error installing ipython
When trying to `conda install anaconda` on a Windows NanoServer container I continually get the below error when it tries to install ipython:
```
ERROR conda.core.link:_execute_actions(330): An error occurred while installing package 'defaults::ipython-5.1.0-py36_0'.
PathNotFoundException()
Attempting to roll back.

PathNotFoundException()
```
This error causes the build of the container to fail (anaconda doesn't get installed). Is there anyway to debug the problem? 

`conda                     4.3.13                   py36_0`

### PR DESCRIPTION
**Pull request by kalefranz**
fix #4703

### CODE REVIEW

File: conda/gateways/disk/create.py [hunk 1]
```diff
@@ -138,8 +138,8 @@ def make_menu(prefix, file_path, remove=False):
         log.warn("Environment name starts with underscore '_'. Skipping menu installation.")
         return
 
-    import menuinst
     try:
+        import menuinst
         menuinst.install(join(prefix, win_path_ok(file_path)), remove, prefix)
     except:
```
Comment (dhirschfeld) [inline conda/gateways/disk/create.py]: Too easy! It's the import of `menuinst` itself which fails on nanoserver so this one-liner should fix it!

**Comment by dhirschfeld** [comment]
```
C:\>echo %USERNAME%
ContainerAdministrator

C:\>python
Python 3.6.0 |Continuum Analytics, Inc.| (default, Dec 23 2016, 11:57:41) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import menuinst
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Miniconda3\lib\site-packages\menuinst\__init__.py", line 23, in <module>
    from .win32 import Menu, ShortCut, dirs
  File "C:\Miniconda3\lib\site-packages\menuinst\win32.py", line 19, in <module>
    dirs = {"system": {"desktop": get_folder_path(FOLDERID.PublicDesktop),
  File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 181, in get_folder_path
    return get_path(folder_id, user)
  File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 164, in get_path
    raise PathNotFoundException()
menuinst.knownfolders.PathNotFoundException
>>>
```

**Comment by dhirschfeld** [comment]
It looks like there's an issue with removing ipython too:
```
<SNIP>
DEBUG conda.core.package_cache:prepare(465): prepared package cache actions:
  cache_actions:

  extract_actions:


DEBUG conda.instructions:execute_instructions(232): executing plan [('PREFIX', 'C:\\Miniconda3'), ('PRINT', 'Unlinking packages ...'), ('PROGRESSIVEFETCHEXTRACT', <cond
', version='5.2.2', build_string='py36_0', build_number=0, with_features_depends=None, base_url=None, platform=None))]
DEBUG conda.instructions:execute_instructions(238):  PREFIX('C:\\Miniconda3')
DEBUG conda.instructions:execute_instructions(238):  PRINT('Unlinking packages ...')
DEBUG conda.instructions:execute_instructions(238):  PROGRESSIVEFETCHEXTRACT(<conda.core.package_cache.ProgressiveFetchExtract object at 0x000002423213DD68>)
DEBUG conda.instructions:execute_instructions(238):  UNLINKLINKTRANSACTION(((Dist(_Dist__initd=True, channel='defaults', dist_name='ipython-5.2.2-py36_0', name='ipython
DEBUG conda.core.link:create_from_dists(117): instantiating UnlinkLinkTransaction with
  target_prefix: C:\Miniconda3
  unlink_dists:
    defaults::ipython-5.2.2-py36_0
  link_dists:


INFO conda.core.link:verify(251): ()
INFO conda.core.link:_execute_actions(306): ===> UNLINKING PACKAGE: defaults::ipython-5.2.2-py36_0 <===
  prefix=C:\Miniconda3

TRACE conda.core.path_actions:execute(628): removing menu for C:\Miniconda3
DEBUG conda.core.link:_execute_actions(323): Error in action #0 for pkg_idx #0 RemoveMenuAction(target_prefix='C:\\Miniconda3', target_short_path='Menu/ipython.json', l

<SNIP>

DEBUG conda.core.link:_execute_actions(324): Traceback (most recent call last):
  File "C:\Miniconda3\lib\site-packages\conda\core\link.py", line 319, in _execute_actions
    action.execute()
  File "C:\Miniconda3\lib\site-packages\conda\core\path_actions.py", line 629, in execute
    make_menu(self.target_prefix, self.target_short_path, remove=True)
  File "C:\Miniconda3\lib\site-packages\conda\gateways\disk\create.py", line 141, in make_menu
    import menuinst
  File "C:\Miniconda3\lib\site-packages\menuinst\__init__.py", line 23, in <module>
    from .win32 import Menu, ShortCut, dirs
  File "C:\Miniconda3\lib\site-packages\menuinst\win32.py", line 19, in <module>
    dirs = {"system": {"desktop": get_folder_path(FOLDERID.PublicDesktop),
  File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 181, in get_folder_path
    return get_path(folder_id, user)
  File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 164, in get_path
    raise PathNotFoundException()
menuinst.knownfolders.PathNotFoundException

ERROR conda.core.link:_execute_actions(330): An error occurred while uninstalling package 'defaults::ipython-5.2.2-py36_0'.
PathNotFoundException()
Attempting to roll back.

INFO conda.core.link:_reverse_actions(349): ===> REVERSING PACKAGE UNLINK: defaults::ipython-5.2.2-py36_0 <===
  prefix=C:\Miniconda3

DEBUG conda.core.link:_reverse_actions(355): reversing pkg_idx #0 from axn_idx #0
TRACE conda.core.path_actions:reverse(632): re-creating menu for C:\Miniconda3
DEBUG conda.core.link:_reverse_actions(364): action.reverse() error in action #0 for pkg_idx #0 RemoveMenuAction(target_prefix='C:\\Miniconda3',

<SNIP>

DEBUG conda.core.link:_reverse_actions(365): Traceback (most recent call last):
  File "C:\Miniconda3\lib\site-packages\conda\core\link.py", line 319, in _execute_actions
    action.execute()
  File "C:\Miniconda3\lib\site-packages\conda\core\path_actions.py", line 629, in execute
    make_menu(self.target_prefix, self.target_short_path, remove=True)
  File "C:\Miniconda3\lib\site-packages\conda\gateways\disk\create.py", line 141, in make_menu
    import menuinst
  File "C:\Miniconda3\lib\site-packages\menuinst\__init__.py", line 23, in <module>
    from .win32 import Menu, ShortCut, dirs
  File "C:\Miniconda3\lib\site-packages\menuinst\win32.py", line 19, in <module>
    dirs = {"system": {"desktop": get_folder_path(FOLDERID.PublicDesktop),
  File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 181, in get_folder_path
    return get_path(folder_id, user)
  File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 164, in get_path
    raise PathNotFoundException()
menuinst.knownfolders.PathNotFoundException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Miniconda3\lib\site-packages\conda\core\link.py", line 361, in _reverse_actions
    action.reverse()
  File "C:\Miniconda3\lib\site-packages\conda\core\path_actions.py", line 633, in reverse
    make_menu(self.target_prefix, self.target_short_path, remove=False)
  File "C:\Miniconda3\lib\site-packages\conda\gateways\disk\create.py", line 141, in make_menu
    import menuinst
  File "C:\Miniconda3\lib\site-packages\menuinst\__init__.py", line 23, in <module>
    from .win32 import Menu, ShortCut, dirs
  File "C:\Miniconda3\lib\site-packages\menuinst\win32.py", line 19, in <module>
    dirs = {"system": {"desktop": get_folder_path(FOLDERID.PublicDesktop),
  File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 181, in get_folder_path
    return get_path(folder_id, user)
  File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 164, in get_path
    raise PathNotFoundException()
menuinst.knownfolders.PathNotFoundException

An unexpected error has occurred.
Please consider posting the following information to the
conda GitHub issue tracker at:

    https://github.com/conda/conda/issues



INFO conda.common.io:captured(59): overtaking stderr and stdout
DEBUG conda.gateways.disk.test:prefix_is_writable(64): testing write access for prefix 'C:\Miniconda3' using path 'C:\Miniconda3\conda-meta\history'
TRACE conda.gateways.disk.test:file_path_is_writable(23): checking path is writable C:\Miniconda3\conda-meta\history
DEBUG conda.gateways.disk.test:prefix_is_writable(64): testing write access for prefix 'C:\Miniconda3' using path 'C:\Miniconda3\conda-meta\history'
TRACE conda.gateways.disk.test:file_path_is_writable(23): checking path is writable C:\Miniconda3\conda-meta\history
INFO conda.common.io:captured(65): stderr and stdout yielding back
Current conda install:

               platform : win-64
          conda version : 4.3.13
       conda is private : False
      conda-env version : 4.3.13
    conda-build version : not installed
         python version : 3.6.0.final.0
       requests version : 2.12.4
       root environment : C:\Miniconda3  (writable)
    default environment : C:\Miniconda3
       envs directories : C:\Miniconda3\envs
                          C:\Users\ContainerAdministrator\conda\conda\envs
                          C:\Users\ContainerAdministrator\.conda\envs
          package cache : C:\Miniconda3\pkgs
                          C:\Users\ContainerAdministrator\conda\conda\pkgs
           channel URLs : https://repo.continuum.io/pkgs/free/win-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/r/win-64
                          https://repo.continuum.io/pkgs/r/noarch
                          https://repo.continuum.io/pkgs/pro/win-64
                          https://repo.continuum.io/pkgs/pro/noarch
                          https://repo.continuum.io/pkgs/msys2/win-64
                          https://repo.continuum.io/pkgs/msys2/noarch
            config file : None
           offline mode : False
             user-agent : conda/4.3.13 requests/2.12.4 CPython/3.6.0 Windows/10 Windows/10.0.14393



`$ C:\Miniconda3\Scripts\conda-script.py remove ipython -v -v -v`




    Traceback (most recent call last):
      File "C:\Miniconda3\lib\site-packages\conda\core\link.py", line 319, in _execute_actions
        action.execute()
      File "C:\Miniconda3\lib\site-packages\conda\core\path_actions.py", line 629, in execute
        make_menu(self.target_prefix, self.target_short_path, remove=True)
      File "C:\Miniconda3\lib\site-packages\conda\gateways\disk\create.py", line 141, in make_menu
        import menuinst
      File "C:\Miniconda3\lib\site-packages\menuinst\__init__.py", line 23, in <module>
        from .win32 import Menu, ShortCut, dirs
      File "C:\Miniconda3\lib\site-packages\menuinst\win32.py", line 19, in <module>
        dirs = {"system": {"desktop": get_folder_path(FOLDERID.PublicDesktop),
      File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 181, in get_folder_path
        return get_path(folder_id, user)
      File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 164, in get_path
        raise PathNotFoundException()
    menuinst.knownfolders.PathNotFoundException

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "C:\Miniconda3\lib\site-packages\conda\core\link.py", line 273, in execute
        pkg_data, actions)
      File "C:\Miniconda3\lib\site-packages\conda\core\link.py", line 337, in _execute_actions
        reverse_excs,
    conda.CondaMultiError:



    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "C:\Miniconda3\lib\site-packages\conda\exceptions.py", line 591, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "C:\Miniconda3\lib\site-packages\conda\cli\main.py", line 134, in _main
        exit_code = args.func(args, p)
      File "C:\Miniconda3\lib\site-packages\conda\cli\main_remove.py", line 205, in execute
        plan.execute_actions(actions, index, verbose=not context.quiet)
      File "C:\Miniconda3\lib\site-packages\conda\plan.py", line 823, in execute_actions
        execute_instructions(plan, index, verbose)
      File "C:\Miniconda3\lib\site-packages\conda\instructions.py", line 247, in execute_instructions
        cmd(state, arg)
      File "C:\Miniconda3\lib\site-packages\conda\instructions.py", line 108, in UNLINKLINKTRANSACTION_CMD
        txn.execute()
      File "C:\Miniconda3\lib\site-packages\conda\core\link.py", line 289, in execute
        rollback_excs,
    conda.CondaMultiError:
```

**Comment by kalefranz** [comment]
oh probably the exact same thing

**Comment by kalefranz** [comment]
Although I think both menu creation and removal use the exact same code path.  So this PR should fix both.

**Comment by dhirschfeld** [comment]
Unfortunately it doesn't appear to - that traceback was from an attempt to `conda remove ipython` after which ipython wasn't removed. I managed to manually install ipython with this patch but not to uninstall it. 
The change is definitely an improvement but I think there must be something else going on...

**Comment by dhirschfeld** [comment]
Ok, likely some issue with my build as the patch didn't make it in :/

**Comment by dhirschfeld** [comment]
Right, I'm definitely not able to remove, even with the patch. Conda log at:
https://gist.github.com/dhirschfeld/be9a83b7e8629797f622fb1e8291c8ee

In the log I first printed the contents of `create.py` on the container to confirm the code was patched:
```
C:\>type "C:\Miniconda3\Lib\site-packages\conda\gateways\disk\create.py"
```
```
def make_menu(prefix, file_path, remove=False):
    """
    Create cross-platform menu items (e.g. Windows Start Menu)

    Passes all menu config files %PREFIX%/Menu/*.json to ``menuinst.install``.
    ``remove=True`` will remove the menu items.
    """
    if not on_win:
        return
    elif basename(prefix).startswith('_'):
        log.warn("Environment name starts with underscore '_'. Skipping menu installation.")
        return

    try:
        import menuinst
        menuinst.install(join(prefix, win_path_ok(file_path)), remove, prefix)
    except:
        pass
        # stdoutlog.error("menuinst Exception:")
        # stdoutlog.error(traceback.format_exc())
```
Trying to remove ipython still raises a `PathNotFoundException` from trying to import menuinst.
```
DEBUG conda.core.link:_execute_actions(324): Traceback (most recent call last):
  File "C:\Miniconda3\lib\site-packages\conda\core\link.py", line 319, in _execute_actions
    action.execute()
  File "C:\Miniconda3\lib\site-packages\conda\core\path_actions.py", line 629, in execute
    make_menu(self.target_prefix, self.target_short_path, remove=True)
  File "C:\Miniconda3\lib\site-packages\conda\gateways\disk\create.py", line 141, in make_menu
    import menuinst
  File "C:\Miniconda3\lib\site-packages\menuinst\__init__.py", line 23, in <module>
    from .win32 import Menu, ShortCut, dirs
  File "C:\Miniconda3\lib\site-packages\menuinst\win32.py", line 19, in <module>
    dirs = {"system": {"desktop": get_folder_path(FOLDERID.PublicDesktop),
  File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 181, in get_folder_path
    return get_path(folder_id, user)
  File "C:\Miniconda3\lib\site-packages\menuinst\knownfolders.py", line 164, in get_path
    raise PathNotFoundException()
menuinst.knownfolders.PathNotFoundException
```
...though the line number suggests it's from the original file?? 😕

**Comment by dhirschfeld** [comment]
I'm not really sure what's going on, but I don't *think* its a conda issue...

**Comment by dhirschfeld** [comment]
..so I'm pretty sure my confusion arose because the conda install operation also updated conda, hence nuking my patched version.

I built the `4.3.x` branch, deployed to nanoserver, tested it and it worked fine so I'll close #4703

**Pull request merged**